benchmark/partial_tweets/nlohmann_json_sax.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#ifdef SIMDJSON_COMPETITION_NLOHMANN_JSON
+
+#include "partial_tweets.h"
+
+namespace partial_tweets {
+
+using json = nlohmann::json;
+
+struct nlohmann_json_sax {
+    using StringType=std::string;
+
+    struct Handler : json::json_sax_t
+    {
+        // 8 keys to parse for each tweet (in order of appearance): "created_at", "id", "text", "in_reply_status_id", "id"(user),
+        // "screen_name"(user), "retweet_count" and "favorite_count".
+        // Assume that the first valid key encountered will be the correct key to parse.
+        // Assume that each tweet/retweet start with a key "metadata" and has a key "retweeted" towards the end
+        // The previous assumption will be used to check for the beginning of a new tweet and the end of a retweet
+        bool keys[8] = { false };   // Stores if previous parsed key was a valid key (in same order as above)
+        bool found[8] = { false }; // Stores if ith key has already been found once
```

Comment:
Instead of `bool [8]`, you might be able to just use an normal word... as a bitset. Let us say that you have `int found = 0`. To set the 6th to 1, you do  `found |= (1<<6)`, to check the value, you do `found &(1<<6)` (returns zero on false, otherwise returns a non-zero value). To set it false, you do `found &= ~(1<<6)`. Doing so would reduce register pressure.

This would be ugly, but you can give the values (1<<0),(1<<1),(1<<2),(1<<3),... names that are pretty...

```C++
enum {
key1 = (1<<0),
key2 = (1<<1),
key3 = (1<<2),
...
};
```

Reply:
I did not look at the assembly of this code, but it is quite possible that each `found[1]` call becomes a load and a compare. A load has a latency of 3 or 4 cycles at a minimum. If you have a sequence of those, the latency piles up.

It is possible that the compiler does something smarter... but I would not trust it to do magic.

In contrast, if you have all the information in a single word then, in the worst case, you may need to load this word once per function call...

Reply:
Ok good idea, I will try it and come back with the results.